source/Phrase.cpp

```diff
@@ -93,6 +93,89 @@ string Phrase::Get() const
 
 
 
+vector<string> Phrase::GetAll() const
+{
+	static auto CombineVectors = [](vector<string> &base, const vector<string> &toAdd) -> void
+	{
+		vector<string> result;
+		result.reserve(base.size() * toAdd.size());
+		for(const string &baseString : base)
+			for(const string &toAppend : toAdd)
+				result.push_back(baseString + toAppend);
+		std::swap(base, result);
+	};
+
+	static auto AppendToVector = [](vector<string> &base, const string &toAppend) -> void
+	{
+		for(string &it : base)
+			it.append(toAppend);
+	};
+
+	static auto PushToVector = [](vector<string> &base, const vector<string> &toAdd) -> void
+	{
+		base.reserve(base.size() + toAdd.size());
+		for(const string &toPush : toAdd)
+			base.push_back(toPush);
+	};
+
+	vector<string> result;
+
+	// Iterate over each sentence in this phrase.
+	// Each sentence is its own set of possible results;
+	// sentences are not concatenated.
+	for(const auto &sentence : sentences)
+	{
+		vector<string> sentenceVector;
+		sentenceVector.emplace_back();
+
+		// Iterate over each part in this sentence.
+		// Sequential parts are concatenated to form a sentence.
+		for(const auto &part : sentence)
+		{
+			vector<string> partVector;
+
+			// Iterate over each choice in this part.
+			// Choices are added onto the previous parts in parallel, not concatenated with each other.
+			for(const auto &choice : part.choices)
+			{
+				vector<string> choiceVector;
+				choiceVector.emplace_back();
+
+				// Iterate over each element in this choice.
+				// The elements in a choice are concatenated.
+				// Each element is either empty, a string, or a pointer to another phrase.
+				for(const auto &element : choice)
+				{
+					// If this element is a string, append it to each of the partial results for this choice.
+					if(!element.first.empty())
+						AppendToVector(choiceVector, element.first);
```

Comment:
This gets used only once, is it really necessary to use a lambda here?

Reply:
Done